File: services/python/shared/telemetry.py
"""
OpenTelemetry Distributed Tracing for Python Services

Provides distributed tracing for:
- FastAPI ML service
- Temporal workflows
- Python microservices

Traces are exported to Jaeger for visualization
"""

import os
import logging
from typing import Optional, Dict, Any

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.semconv.resource import ResourceAttributes
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentation
from opentelemetry.instrumentation.requests import RequestsInstrumentation
from opentelemetry.instrumentation.logging import LoggingInstrumentation

logger = logging.getLogger(__name__)

class TelemetryConfig:
    """Configuration for OpenTelemetry"""
    
    def __init__(
        self,
        service_name: str,
        service_version: str = "1.0.0",
        jaeger_endpoint: str = "http://localhost:14268/api/traces",
        enabled: bool = True
    ):
        self.service_name = service_name
        self.service_version = service_version
        self.jaeger_endpoint = jaeger_endpoint
        self.enabled = enabled


def init_telemetry(config: TelemetryConfig) -> Optional[TracerProvider]:
    """
    Initialize OpenTelemetry with Jaeger exporter
    
    Args:
        config: TelemetryConfig instance
        
    Returns:
        TracerProvider instance or None if disabled
    """
    if not config.enabled:
        logger.info("[Telemetry] Tracing disabled")
        return None
    
    try:
        # Create resource with service information
        resource = Resource.create({
            ResourceAttributes.SERVICE_NAME: config.service_name,
            ResourceAttributes.SERVICE_VERSION: config.service_version,
            ResourceAttributes.DEPLOYMENT_ENVIRONMENT: os.getenv("NODE_ENV", "development"),
        })
        
        # Create OTLP exporter for Jaeger
        exporter = OTLPSpanExporter(
            endpoint=config.jaeger_endpoint,
            headers={},
        )
        
        # Create trace provider
        provider = TracerProvider(resource=resource)
        processor = BatchSpanProcessor(exporter)
        provider.add_span_processor(processor)
        
        # Set global trace provider
        trace.set_tracer_provider(provider)
        
        # Auto-instrument common libraries
        RequestsInstrumentation().instrument()
        LoggingInstrumentation().instrument()
        
        logger.info(f"[Telemetry] OpenTelemetry initialized for {config.service_name}")
        logger.info(f"[Telemetry] Exporting traces to: {config.jaeger_endpoint}")
        
        return provider
        
    except Exception as e:
        logger.error(f"[Telemetry] Failed to initialize OpenTelemetry: {e}")
        return None


def instrument_fastapi(app):
    """
    Instrument FastAPI application with OpenTelemetry
    
    Args:
        app: FastAPI application instance
    """
    try:
        FastAPIInstrumentation.instrument_app(app)
        logger.info("[Telemetry] FastAPI instrumented")
    except Exception as e:
        logger.error(f"[Telemetry] Failed to instrument FastAPI: {e}")


def get_tracer(service_name: str) -> trace.Tracer:
    """
    Get a tracer for the given service
    
    Args:
        service_name: Name of the service
        
    Returns:
        Tracer instance
    """
    return trace.get_tracer(service_name)


def start_span(service_name: str, span_name: str, attributes: Optional[Dict[str, Any]] = None):
    """
    Start a new span with the given name
    
    Args:
        service_name: Name of the service
        span_name: Name of the span
        attributes: Optional attributes to add to the span
        
    Returns:
        Span context manager
    """
    tracer = get_tracer(service_name)
    span = tracer.start_span(span_name)
    
    if attributes:
        for key, value in attributes.items():
            span.set_attribute(key, value)
    
    return span


def add_span_attributes(span: trace.Span, attributes: Dict[str, Any]):
    """
    Add attributes to the current span
    
    Args:
        span: Span instance
        attributes: Dictionary of attributes to add
    """
    for key, value in attributes.items():
        span.set_attribute(key, value)


def add_span_event(span: trace.Span, name: str, attributes: Optional[Dict[str, Any]] = None):
    """
    Add an event to the current span
    
    Args:
        span: Span instance
        name: Event name
        attributes: Optional event attributes
    """
    if attributes:
        span.add_event(name, attributes)
    else:
        span.add_event(name)


def record_error(span: trace.Span, error: Exception):
    """
    Record an error on the span
    
    Args:
        span: Span instance
        error: Exception to record
    """
    span.record_exception(error)
    span.set_attribute("error", True)


def shutdown_telemetry(provider: Optional[TracerProvider]):
    """
    Shutdown the telemetry provider
    
    Args:
        provider: TracerProvider instance
    """
    if provider:
        try:
            provider.shutdown()
            logger.info("[Telemetry] Tracer provider shut down successfully")
        except Exception as e:
            logger.error(f"[Telemetry] Error shutting down tracer provider: {e}")

File: services/go/shared/telemetry.go
package shared

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
	"go.opentelemetry.io/otel/trace"
)

// TelemetryConfig holds OpenTelemetry configuration
type TelemetryConfig struct {
	ServiceName    string
	ServiceVersion string
	JaegerEndpoint string
	Enabled        bool
}

// InitTelemetry initializes OpenTelemetry with Jaeger exporter
func InitTelemetry(config TelemetryConfig) (func(context.Context) error, error) {
	if !config.Enabled {
		log.Println("[Telemetry] Tracing disabled")
		return func(ctx context.Context) error { return nil }, nil
	}

	// Default Jaeger endpoint
	if config.JaegerEndpoint == "" {
		config.JaegerEndpoint = "http://localhost:14268/api/traces"
	}

	// Create OTLP exporter
	ctx := context.Background()
	exporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint("localhost:14268"),
		otlptracehttp.WithInsecure(),
		otlptracehttp.WithURLPath("/api/traces"),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create OTLP exporter: %w", err)
	}

	// Create resource with service information
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceNameKey.String(config.ServiceName),
			semconv.ServiceVersionKey.String(config.ServiceVersion),
			attribute.String("environment", getEnv("NODE_ENV", "development")),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create resource: %w", err)
	}

	// Create trace provider
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)

	// Set global trace provider
	otel.SetTracerProvider(tp)

	// Set global propagator for context propagation
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	log.Printf("[Telemetry] OpenTelemetry initialized for %s", config.ServiceName)
	log.Printf("[Telemetry] Exporting traces to: %s", config.JaegerEndpoint)

	// Return shutdown function
	shutdown := func(ctx context.Context) error {
		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
		defer cancel()
		
		if err := tp.Shutdown(ctx); err != nil {
			return fmt.Errorf("failed to shutdown tracer provider: %w", err)
		}
		
		log.Println("[Telemetry] Tracer provider shut down successfully")
		return nil
	}

	return shutdown, nil
}

// GetTracer returns a tracer for the given service
func GetTracer(serviceName string) trace.Tracer {
	return otel.Tracer(serviceName)
}

// StartSpan creates a new span with the given name
func StartSpan(ctx context.Context, serviceName, spanName string) (context.Context, trace.Span) {
	tracer := GetTracer(serviceName)
	return tracer.Start(ctx, spanName)
}

// AddSpanAttributes adds attributes to the current span
func AddSpanAttributes(span trace.Span, attrs map[string]interface{}) {
	for key, value := range attrs {
		switch v := value.(type) {
		case string:
			span.SetAttributes(attribute.String(key, v))
		case int:
			span.SetAttributes(attribute.Int(key, v))
		case int64:
			span.SetAttributes(attribute.Int64(key, v))
		case float64:
			span.SetAttributes(attribute.Float64(key, v))
		case bool:
			span.SetAttributes(attribute.Bool(key, v))
		default:
			span.SetAttributes(attribute.String(key, fmt.Sprintf("%v", v)))
		}
	}
}

// AddSpanEvent adds an event to the current span
func AddSpanEvent(span trace.Span, name string, attrs map[string]interface{}) {
	var attributes []attribute.KeyValue
	for key, value := range attrs {
		switch v := value.(type) {
		case string:
			attributes = append(attributes, attribute.String(key, v))
		case int:
			attributes = append(attributes, attribute.Int(key, v))
		case int64:
			attributes = append(attributes, attribute.Int64(key, v))
		case float64:
			attributes = append(attributes, attribute.Float64(key, v))
		case bool:
			attributes = append(attributes, attribute.Bool(key, v))
		default:
			attributes = append(attributes, attribute.String(key, fmt.Sprintf("%v", v)))
		}
	}
	span.AddEvent(name, trace.WithAttributes(attributes...))
}

// RecordError records an error on the span
func RecordError(span trace.Span, err error) {
	if err != nil {
		span.RecordError(err)
		span.SetAttributes(attribute.Bool("error", true))
	}
}

// Helper function to get environment variable with default
func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}
